File: helpers/ecma/match.mjs
#!/usr/bin/env node
/** Replay: node match.mjs [--batch] <pattern> <flags> — reads stdin.
 *
 * Single mode (default): stdin is ONE witness (raw bytes); exit 0 on test().
 * Batch mode: `node match.mjs --batch <pattern> <flags>` — stdin is the
 * NUL-framed batch stream; emits one `<index>:<verdict>` line per witness on
 * stdout (0 = rejected, 1 = accepted). The `--batch` flag form (not a
 * positional `batch`) keeps a pattern literally named "batch" unambiguous.
 *
 * Exit codes (helper contract): 0 = accepted, 1 = rejected, 2 = compile /
 * engine error. Batch framing: each witness is escaped byte-wise (0x00 →
 * `\0`, `\` → `\\`) before being NUL-delimited, so witnesses containing NUL
 * round-trip exactly across the NUL delimiter.
 *
 * Pattern/flags are CLI args to this ground-truth replay harness. Escaping
 * them would change language membership (the pattern *is* the SUT). Compile
 * failures exit 2; hang/ReDoS is bounded by the caller subprocess timeout.
 * Not an untrusted-service injection boundary — see docs/SECURITY-AUDIT.md. */
const batch = process.argv[2] === "--batch";
const pattern = batch ? process.argv[3] ?? "" : process.argv[2] ?? "";
const flags = batch ? process.argv[4] ?? "" : process.argv[3] ?? "";
let re;
try {
  // codeql[js/regex-injection] Harness argv is the pattern under test; escaping would change fuzzing semantics.
  re = new RegExp(pattern, flags);
} catch (e) {
  console.error(e);
  process.exit(2);
}

async function readStdin() {
  const chunks = [];
  for await (const c of process.stdin) chunks.push(c);
  return Buffer.concat(chunks);
}

/** Split a Buffer on raw NUL bytes (the escaped frames contain none). */
function splitNul(buf) {
  const parts = [];
  let start = 0;
  for (let i = 0; i < buf.length; i++) {
    if (buf[i] === 0) {
      parts.push(buf.subarray(start, i));
      start = i + 1;
    }
  }
  parts.push(buf.subarray(start));
  return parts;
}

/** Decode one NUL-framed witness: 0x00 → `\0`, `\` → `\\` escape layer. */
function decodeFrame(buf) {
  const out = [];
  for (let i = 0; i < buf.length; i++) {
    const b = buf[i];
    if (b === 0x5c) {
      const n = buf[i + 1];
      if (n === 0x5c) {
        out.push(0x5c);
        i++;
      } else if (n === 0x30) {
        out.push(0x00);
        i++;
      } else {
        out.push(b);
      }
    } else {
      out.push(b);
    }
  }
  return Buffer.from(out).toString("utf8");
}

async function main() {
  const raw = await readStdin();
  if (!batch) {
    const s = raw.toString("utf8");
    process.exit(re.test(s) ? 0 : 1);
  }
  const frames = splitNul(raw);
  if (frames.length && frames[frames.length - 1].length === 0) frames.pop();
  for (let i = 0; i < frames.length; i++) {
    const s = decodeFrame(frames[i]);
    console.log(`${i}:${re.test(s) ? 1 : 0}`);
  }
}

main();

File: helpers/python/match.py
#!/usr/bin/env python3
"""Python `re` ground-truth helper — timed subprocess runner owned by the
ground-truth adapter (``regexproof/groundtruth/adapters.py``).

The adapter routes py_re replay through this helper so ``timeout_s`` is ALWAYS
honored: a catastrophic pattern would block the gate indefinitely in-process
(P1 finding 3), here the subprocess timeout bounds it. ``re.fullmatch`` /
``re.match`` / ``re.search`` semantics are honored directly — no pattern
wrapping (for the subprocess dialects whose helpers are bare search runners the
adapter wraps the pattern instead; py_re's whole-string/prefix/search intent is
the engine's native call).

Usage:
  match.py match <call_kind> <pattern> <flags>   # stdin → verdict on stdout
  match.py batch <call_kind> <pattern> <flags>   # NUL-framed witnesses

Exit codes (helper contract shared with the other match helpers):
  0 = accepted, 1 = rejected, 2 = engine/compile error, 3 = runner unavailable.

Batch framing: witnesses are NUL-delimited on stdin; each witness is escaped
byte-wise (0x00 → ``\\0``, ``\\`` → ``\\\\``) before joining so witnesses
containing NUL round-trip exactly (raw NUL never appears inside an escaped
segment, so NUL is an unambiguous delimiter). The helper emits one
``<index>:<verdict>`` line per witness on stdout (0 rejected / 1 accepted).
"""

from __future__ import annotations

import re
import sys

_FLAG_BITS = {
    "i": re.IGNORECASE,
    "m": re.MULTILINE,
    "s": re.DOTALL,
    "x": re.VERBOSE,
    "a": re.ASCII,
    "u": re.UNICODE,
}


def _compile(pattern: str, flags: str):
    bits = 0
    for ch in flags or "":
        if ch not in _FLAG_BITS:
            raise ValueError(f"unknown py_re flag {ch!r}")
        bits |= _FLAG_BITS[ch]
    return re.compile(pattern, bits)


def _match(rx, call_kind: str, data: str) -> bool:
    if call_kind == "fullmatch":
        return rx.fullmatch(data) is not None
    if call_kind == "match":
        return rx.match(data) is not None
    return rx.search(data) is not None  # search / exec


def _decode_frame(raw: bytes) -> str:
    out = bytearray()
    i = 0
    n = len(raw)
    while i < n:
        b = raw[i]
        if b == 0x5C:  # backslash escape
            nxt = raw[i + 1] if i + 1 < n else None
            if nxt == 0x5C:
                out.append(0x5C)
                i += 2
                continue
            if nxt == 0x30:  # '0'
                out.append(0x00)
                i += 2
                continue
            out.append(b)
            i += 1
            continue
        out.append(b)
        i += 1
    return bytes(out).decode("utf-8")


def _read_frames() -> list[str]:
    raw = sys.stdin.buffer.read()
    frames = raw.split(b"\x00")
    if frames and frames[-1] == b"":
        frames.pop()
    return [_decode_frame(f) for f in frames]


def cmd_match(call_kind: str, pattern: str, flags: str) -> int:
    try:
        data = sys.stdin.buffer.read().decode("utf-8")
    except UnicodeDecodeError as exc:
        print(f"engine-error: {exc}", file=sys.stderr)
        return 2
    try:
        rx = _compile(pattern, flags)
        accepted = _match(rx, call_kind, data)
    except (re.error, ValueError) as exc:
        print(f"engine-error: {exc}", file=sys.stderr)
        return 2
    print("accepted" if accepted else "rejected")
    return 0 if accepted else 1


def cmd_batch(call_kind: str, pattern: str, flags: str) -> int:
    witnesses = _read_frames()
    try:
        rx = _compile(pattern, flags)
    except (re.error, ValueError) as exc:
        print(f"engine-error: {exc}", file=sys.stderr)
        return 2
    for i, w in enumerate(witnesses):
        try:
            accepted = _match(rx, call_kind, w)
        except re.error as exc:  # engine error mid-batch aborts the session
            print(f"engine-error: {exc}", file=sys.stderr)
            return 2
        print(f"{i}:{1 if accepted else 0}")
    return 0


def main(argv) -> int:
    if len(argv) < 3 or argv[0] not in ("match", "--batch"):
        print(
            "usage: match.py match|batch <call_kind> <pattern> [flags]",
            file=sys.stderr,
        )
        return 2
    cmd, call_kind, pattern = argv[0], argv[1], argv[2]
    flags = argv[3] if len(argv) > 3 else ""
    if cmd == "match":
        return cmd_match(call_kind, pattern, flags)
    return cmd_batch(call_kind, pattern, flags)


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
